File: runtime/release_readiness.ts
import { collectProviderStatusWithOptions } from "./provider_bootstrap.ts";

function git(projectDir: string, args: string[]): string {
  const proc = Bun.spawnSync({
    cmd: ["git", ...args],
    cwd: projectDir,
    stdout: "pipe",
    stderr: "pipe"
  });
  return proc.exitCode === 0 ? proc.stdout.toString().trim() : "";
}

export function collectReleaseReadiness(projectDir: string) {
  const providerStatus = collectProviderStatusWithOptions(projectDir, {
    includeSmoke: true,
    smokeHostMode: "claude_dispatch"
  });
  const providers = providerStatus.providers as Array<Record<string, unknown>>;
  const blocked = providers
    .filter((entry) => String(entry.parity_state) === "blocked")
    .map((entry) => String(entry.provider));
  const native_ready = providers
    .filter((entry) => Boolean(entry.native_ready))
    .map((entry) => String(entry.provider));
  const blockers = providers.flatMap((entry) => [
    ...(Array.isArray(entry.local_steps) ? (entry.local_steps as string[]).map((step) => `${entry.provider}: ${step}`) : []),
    ...(Array.isArray(entry.provider_steps) ? (entry.provider_steps as string[]).map((step) => `${entry.provider}: ${step}`) : [])
  ]);
  const statusLines = git(projectDir, ["status", "--short"])
    .split(/\r?\n/)
    .filter(Boolean);
  return {
    schema: "OmgReleaseReadiness",
    status: "ok",
    git: {
      branch: git(projectDir, ["branch", "--show-current"]),
      dirty: statusLines.length > 0,
      status_lines: statusLines
    },
    providers: {
      blocked,
      native_ready,
      matrix: providerStatus
    },
    blockers,
    ready_for_release: blocked.length === 0 && statusLines.length === 0
  };
}

File: runtime/release_readiness.py
"""Release-readiness summary for provider/runtime/git state."""
from __future__ import annotations

import subprocess
from typing import Any

from runtime.provider_bootstrap import collect_provider_status_with_options


def _git_branch(project_dir: str) -> str:
    proc = subprocess.run(
        ["git", "branch", "--show-current"],
        cwd=project_dir,
        capture_output=True,
        text=True,
        check=False,
        timeout=10,
    )
    return proc.stdout.strip()


def _git_status_lines(project_dir: str) -> list[str]:
    proc = subprocess.run(
        ["git", "status", "--short"],
        cwd=project_dir,
        capture_output=True,
        text=True,
        check=False,
        timeout=10,
    )
    return [line.rstrip() for line in proc.stdout.splitlines() if line.strip()]


def collect_release_readiness(project_dir: str) -> dict[str, Any]:
    provider_status = collect_provider_status_with_options(
        project_dir,
        include_smoke=True,
        smoke_host_mode="claude_dispatch",
    )
    providers = provider_status.get("providers", [])
    blocked = [
        str(entry.get("provider", ""))
        for entry in providers
        if isinstance(entry, dict) and str(entry.get("parity_state", "")) == "blocked"
    ]
    native_ready = [
        str(entry.get("provider", ""))
        for entry in providers
        if isinstance(entry, dict) and bool(entry.get("native_ready"))
    ]
    blockers: list[str] = []
    for entry in providers:
        if not isinstance(entry, dict):
            continue
        provider_name = str(entry.get("provider", ""))
        for step in entry.get("local_steps", []):
            blockers.append(f"{provider_name}: {step}")
        for step in entry.get("provider_steps", []):
            blockers.append(f"{provider_name}: {step}")

    git_status = _git_status_lines(project_dir)
    return {
        "schema": "OmgReleaseReadiness",
        "status": "ok",
        "git": {
            "branch": _git_branch(project_dir),
            "dirty": bool(git_status),
            "status_lines": git_status,
        },
        "providers": {
            "blocked": blocked,
            "native_ready": native_ready,
            "matrix": provider_status,
        },
        "blockers": blockers,
        "ready_for_release": not blocked and not git_status,
    }
